merkledag/merkledag_cache.go

```diff
@@ -0,0 +1,69 @@
+// package merkledag implements the ipfs Merkle DAG datastructures.
+package merkledag
+
+import (
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	"gx/ipfs/QmVYxfoJQiZijTgPNHCHgHELvQpbsJNTg6Crmc3dQkj3yy/golang-lru"
+	"gx/ipfs/QmZy2y8t9zQH2a1b8q2ZSLKp17ATuJoCNxxyMFG5qFExpt/go-net/context"
+)
+
+func NewDAGCache(ds DAGService, size int) (DAGService, error) {
+	cache, err := lru.New(size)
+	if err != err {
+		return nil, err
+	}
+	return &dagCache{ds, cache}, nil
+}
+
+type dagCache struct {
+	DAGService
+	cache *lru.Cache
+}
+
+func CloneNode(nd * Node) *Node {
```

Comment:
That is an overkil.  A shallow copy is all that is needed.  I don't think any sort of copying should be necessary as it is in general bad practice to modify an object returned from the cache.